### PR TITLE
Ask to Push Key After Generating a New Pair 

### LINKED ISSUE #2907: Generating a New Keypair Should Ask to push it to the Key Server
Feature requested by @gmkurtzer.

<br>

## Version of Singularity:

```
3.1.0-317.gf7d868b5
```

### Expected behavior

Generating a new key should ask you to push it to the keystore:

```
$ singularity key newpair 
Enter your name (e.g., John Doe) : WestleyK
Enter your email address (e.g., john.doe@example.com) : westley@sylabs.io
Enter optional comment (e.g., development keys) : test
Generating Entity and OpenPGP Key Pair... Done
Enter encryption passphrase : 
Would you like to push you key to the library [Y,n] : y
Done.
```

### Actual behavior

```
$ singularity key newpair 
Enter your name (e.g., John Doe) : WestleyK
Enter your email address (e.g., john.doe@example.com) : westley@sylabs.io
Enter optional comment (e.g., development keys) : test
Generating Entity and OpenPGP Key Pair... Done
Enter encryption passphrase : 
```

### Steps to reproduce behavior

```
$ singularity key newpair 
```

### PR DESCRIPTION
**Pull request by WestleyK**
## Description of the Pull Request (PR):

When you generate a newpair, you will be asked if you want to push it to the keystore.

#### Example:

```go
$ singularity key newpair 
WARNING: sylabs-token files are deprecated. Use 'singularity remote' to manage remote endpoints and tokens.
Enter your name (e.g., John Doe) : westley 
Enter your email address (e.g., john.doe@example.com) : westley@sylabs.io
Enter optional comment (e.g., development keys) : test key; do not use
Enter encryption passphrase : 
Retype your passphrase : 
Generating Entity and OpenPGP Key Pair...done
Would you like to push it to the keystore? [Y,n] : 
Key successfully pushed to: https://keys.sylabs.io
Done.
```

***NOTE:** This PR should be merged after [Confirm Key Generate Password #2886](https://github.com/sylabs/singularity/pull/2886), to minimize conflict.*

## This fixes or addresses the following GitHub issues:

- Fixes #2907 

<br>

### CODE REVIEW

**Review by jchavezlavalle — APPROVED** [review]
lgtm :+1:

File: cmd/internal/cli/key_newpair.go [hunk 1]
```diff
@@ -31,3 +35,9 @@ var KeyNewPairCmd = &cobra.Command{
 	Long:    docs.KeyNewPairLong,
 	Example: docs.KeyNewPairExample,
 }
+
+// doKeyNewpairCmd is the 'singularity key newpair' command
+func doKeyNewpairCmd(keyServerURI, authToken string) error {
```
Comment (ikaneshiro) [inline cmd/internal/cli/key_newpair.go]: Why have a helper function when its passing parameters to `GenKeyPair()` and returning the result without any additional logic?

File: cmd/internal/cli/key_newpair.go [hunk 2]
```diff
@@ -31,3 +35,9 @@ var KeyNewPairCmd = &cobra.Command{
...
+	_, err := sypgp.GenKeyPair(keyServerURL, authToken)
```
Comment (ikaneshiro) [inline cmd/internal/cli/key_newpair.go]: Was `keyServerURI` or `keyServerURL` indended? The latter being a global variable in this package.

**Review by ikaneshiro — CHANGES_REQUESTED** [review]
Couple small things

File: cmd/internal/cli/key_newpair.go (hunk 1 above)
Comment (WestleyK) [reply cmd/internal/cli/key_newpair.go]: This function is unnecessary, it only *was* there for consistent with other file. *(now that I'm looking at it, its probably better to remove this function)*

Comment (ikaneshiro) [reply cmd/internal/cli/key_newpair.go]: Agreed :+1:

File: cmd/internal/cli/key_newpair.go (hunk 2 above)
Comment (WestleyK) [reply cmd/internal/cli/key_newpair.go]: I think the `keyServerURL` is the global variable in this package... Should it be renamed to `keyServerURI`?

Comment (ikaneshiro) [reply cmd/internal/cli/key_newpair.go]: I think either is acceptable, though `URL` is probably more accurate in this case

File: pkg/sypgp/sypgp.go [hunk 3]
```diff
@@ -372,18 +373,37 @@ func GenKeyPair() (entity *openpgp.Entity, err error) {
 		return
 	}
 
-	fmt.Print("Generating Entity and OpenPGP Key Pair... ")
-	entity, err = openpgp.NewEntity(name, comment, email, conf)
+	pass, err := AskQuestionNoEcho("Enter encryption passphrase : ")
```
Comment (mem) [inline pkg/sypgp/sypgp.go]: Can we please move the asking to a function and handle the retries using that?

Something like this:

```
var errPassphraseMismatch = errors.New("passphrases do not match")
var errTooManyRetries = errors.New("too many retries while getting a passphrase")

func getPassphrase(retries int) (string, error) {
    ask := func() (string, error) {
        pass1, err := AskQuestinNoEcho("...")
        if err != nil {
            return "", err
        }

        pass2, err := AskQuestinNoEcho("...")
        if err != nil {
            return "", err
        }

        if pass1 != pass2 {
            return "", errPassphraseMismatch
        }

        return pass1, nil
    }

    for i := 0; i < retries; i++ {
        switch passphrase, err := ask(); err {
        case nil:
            // we got it!
           return passphrase, nil
        case errPassphraseMismatch:
            // retry
        default:
            // something else went wrong, bail out
            return "", err
        }
    }

    return "", errTooManyRetries
}
```

and then here you can do:

```
    passphrase, err := getPassphrase(3)
    if err != nil {
        return ...
    } 
```

File: cmd/internal/cli/key_newpair.go (hunk 2 above)
Comment (mem) [reply cmd/internal/cli/key_newpair.go]: this should be keyServerURI, the argument to the function. The argument to the function comes from keyServerURL (L#27 above), but for testability's sake this should not use globals.

File: pkg/sypgp/sypgp.go (hunk 3 above)
Comment (WestleyK) [reply pkg/sypgp/sypgp.go]: Resolved by [`
35271f8`](https://github.com/sylabs/singularity/pull/2886/commits/35271f8dfd516f0871a2b22f911a1a9347eeac3d).

**Review by mem — APPROVED** [review]
LGTM, thanks!

**Review by ikaneshiro — APPROVED** [review]

**Pull request merged**